### PR TITLE
feat(rest): app.route() and app.api()

### PR DESCRIPTION
**Pull request by bajtos**
Enhance `RestApplication` class to provide shortcuts for registering routes and configuring the master OpenAPI spec.

See #846 and #955.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] Related API Documentation was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `packages/example-*` were updated

### CODE REVIEW

File: packages/rest/test/acceptance/routing/routing.acceptance.ts [hunk 1]
```diff
@@ -487,6 +488,83 @@ describe('Routing', () => {
     await client.get('/greet?name=world').expect(200, 'hello world');
   });
 
+  describe('RestApplication', () => {
+    it('supports function-based routes', async () => {
```
Comment (bajtos) [inline packages/rest/test/acceptance/routing/routing.acceptance.ts]: This is a copy of an existing test which is using `server.route`.

**Comment by bajtos** [comment]
The tests are a copy and paste of existing tests using `server.*` instead of `app.*` APIs.

**Comment by raymondfeng** [comment]
Travis fails with the following error:
```
1) Routing
       RestApplication
         supports controller routes declared via app.api():
     Error: expected 200 "OK", got 404 "Not Found"
      at Test._assertStatus (packages/testlab/node_modules/supertest/lib/test.js:266:12)
      at Test._assertFunction (packages/testlab/node_modules/supertest/lib/test.js:281:11)
      at Test.assert (packages/testlab/node_modules/supertest/lib/test.js:171:18)
      at Server.assert (packages/testlab/node_modules/supertest/lib/test.js:131:12)
      at emitCloseNT (net.js:1655:8)
      at _combinedTickCallback (internal/process/next_tick.js:135:11)
      at process._tickCallback (internal/process/next_tick.js:180:9)
```

Please fix.

**Comment by bajtos** [comment]
> tests were missing for sequence and handler functions.

`app.handler` is tested here:
https://github.com/strongloop/loopback-next/blob/baf9928613ccde82c02abbc8249f4163e6a5570b/packages/rest/test/acceptance/bootstrapping/rest.acceptance.ts#L55-L60

I have added a new test for `app.sequence`, see e923d875e3.

Now I need to figure out why the tests are failing.

**Comment by bajtos** [comment]
It turns out `restApp.sequence` was broken, I fixed it in 7cc0777 (feel free to review).

What's remaining: fix `restApp.api()` test(s).

**Comment by bajtos** [comment]
All should be good now.

@raymondfeng @shimks @kjdelisle please review

> Thanks for essentially finishing my task here

@shimks I think we need to update any examples in our documentation on loopback.io to use these new APIs. I'll post a comment in the issue.

**Pull request merged**